### ISSUE TITLE
email_src_async may redirect to https://mail-attachment.googleusercontent.com, which is blocked by cors

### ISSUE THREAD

**Issue by mgild** [KartikTalwar/gmail.js#384]
Use case:

> emails_on_page.forEach(function(email) {
        var email_id = email["id"];
        // Encoding may go wrong sometimes.  May have to parse the binary.
        // Read email.js docs about email_source_async
        gmail.get.email_source_async(email_id, function (raw_email) {
            var email_headers = extractHeaderBlock(raw_email, false, true);
            console.log("Email ID: " + email_id);
            console.log(email_headers.headers);
        }, function(e){console.error(email_id + " error occurred: " + e)}, false);

**Comment by josteink** [comment]
Thanks for your report, and thanks for including a fairly complete POC code-sample!

From what I can tell, you're trying to get the MIME-source *all* emails on a page, and process all of them in order. But that's not what you're doing here, and this is why things go wrong :)

When we're fetching the MIME-source from Gmail, we're asking for a privileged resource, only available to those who should have access. And Google does a very good job protecting it. I've spent *weeks* hunting down this bug, so I know pretty much what's going on.

Consider the following example, where we fetch the MIME-data for *one* email:

* We request a single HTTPS resource on `mail.google.com`.
* This gets us some cookies, and a redirect to `mail-attachment.googleusercontent.com`. This resource is on a different domain, and will only be accessible with the proper cookies!
* Immediately when we have retrieved this resource, the browser is asked to clear our cookies for this domain.

If we were to try to request another resource form `mail-attachment.googleusercontent.com` at this point, we'd be redirected to `mail.google.com` due to lack of authentication. But `mail.google.com` has no CORS-headers for `mail-attachment.googleusercontent.com`, so any such XHR will fail.

Now... Let's look at what you're doing:

* You take all emails on page.
* Quickly, for each email, you queue an async download which will cause the same as above to happen, but in parallell!
* This means that as soon as one email is downloaded, your cookies will be cleared and all subsequent requests will fail! And that's a not something we can fix.

To make this thing work, I suspect you will need to write something along the following lines:

````javascript
var emails_to_dl = emails_on_page;
var emails_dled = [];

function download_emails(callback) {
    var email_id = emails_to_dl.pop();
    gmail.get.email_source_async(email_id, function(res) {
        emails_dled.push(res);
        if (emails_to_dl.length === 0) {
              callback(emails_dled);
        } else {
              download_emails(callback);
        }
    });
}

download_emails(function(emails) {
    // do stuff here
});
````

Something along those lines.

That will ensure you only download one email at a time, instead of doing it in parallell.

Hope this helps, and hope you find a solution that works for you.
